### ISSUE TITLE
nextcloud 11 beta 2 passwords list empty "unexpected token f in JSON (...)"

### ISSUE THREAD

**Issue by pafcioooo** [fcturner/passwords#296]
After upgrading from NextCloud 10 to Nextcloud 11 beta 2, Passwords application does not show any active passwords.
It loads application gui, but while loading list of passwords it displays two windows with errors and after that active passwords list is empty, and all entries are located in recycle bin.
See the error screenshots below (the GUI is in Polish language):
![image](https://cloud.githubusercontent.com/assets/891504/20732631/f713a01c-b68f-11e6-990c-4211c369e4b4.png)
![image](https://cloud.githubusercontent.com/assets/891504/20732670/18ec3e42-b690-11e6-9665-ab0e64307bf9.png)

**Comment by pafcioooo** [comment]
I have edited issue description according to a new finding that all passwords are in recycle bin list.

**Comment by zguithues** [comment]
+1 same here with Nextcloud 11 beta 2

**Comment by fcturner** [comment]
Thanks for reporting! I should have some time next week to look at it. 

You're all making this app great by supplying me with smaller and bigger issues! Great contribution, thanks! :+1:

**Comment by pafcioooo** [comment]
With NC  11.0 RC1 issue still exists.

**Comment by pafcioooo** [comment]
With NC 11.0.0 official release bug still exists:-(

**Comment by fcturner** [comment]
That's a bummer :disappointed: 

That Polish language surely looks exotic! I think something might have to do with that. I'll switch to Polish and do some research :+1:

**Comment by pafcioooo** [comment]
Well I do not use Polish language in password entries. After switching to English user interface the problem still exists.

**Comment by Pyhrra** [comment]
I am currently getting the issue with "unexpected token f in JSON (...)" , currently using NC 11 with 20 Beta6. 
Recycling is all present but primary entries do not decrypt or present themselves.

I have reverted all to NC10 with Beta6 and all is working.

**Comment by AlanFordddd** [comment]
Same error here.

**Comment by nbourdais** [comment]
Same here, but I've got message for each nextcloud user.
Could it be linked with issue #280 ?

**Comment by Happyfeet01** [comment]
i have the same Issue with Nc 11

SyntaxError: JSON.parse: expected double-quoted property name at line 1 column 40 of the JSON data ....

**Comment by lucashalbert** [comment]
+1 same here with English Nextcloud 11.0.0 (stable). All active passwords ended up in the Trash Bin after upgrade.

Great job on this app fcturner!

**Comment by fcturner** [comment]
@lucashalbert Please don't do this. I'm just a hobby developer trying to create and maintain a nice app. I never asked you to use this app on the latest versions of NextCloud that haven't been tested with my app. I like this community but I think your comment is just rude, I'm sorry. 

This app works perfectly on ownCloud. I really can't follow the NextCloud developers. I constantly need to compensate for their updates without any documentation. That's hard, sadly to say. Holidays are coming, time is freeing up and I'm still really enthusiastic about this app, as I know that others are!

**Comment by fcturner** [comment]
Wait, you WERE cynical, weren't you? About the 'great job' thing? I apologise otherwise, really got you wrong then.

**Comment by lucashalbert** [comment]
I was not being cynical in any way @fcturner. I was merely trying to give you a compliment. If I didn't like the app I wouldn't use it.

Thanks for your hard work and Happy Holidays!

**Comment by fcturner** [comment]
Ahh, thank you!! Excuse me, really. Like I said, really got you wrong :wink:

Thanks again!!

**Comment by ezraholm50** [comment]
Same issue here, getting the popup for about 5 users. Passwords are still visible.

When trying to reshare a password the app cant find other users to share with. Perhaps you've already figured this out. Thanks for this awesome app.

**Comment by seraphyn** [comment]
Problem still there, but I can live with it atm.
thanks for the great app and merry christmas to you @fcturner

**Comment by Luticus** [comment]
I have the issue and unfortunately it doesn't seem to be printing anything in the logs.  for me though all the tables/passwords draw and work fine (chrome).  If I could just suppress the error message even it would be fine for now.  :)  The specific error I get is "syntaxerror unexpected token f in json at postion 43 (row 7)".  then it shows the same error over and over with different position/row numbers.  after it shows the last one, the errors go away and the interface looks and seems to work as normal.  I really hate that owncloud forked because now exactly half my apps have gone to one and half have gone to the other and I'm always fighting to make them all work.  You're password app really is the best now though, can't live without it.  Hoping the fix isn't too tough to work out.  If there's any logs/help whatever that I can provide to make it easier I'd be glad to do so.

**Comment by tapir** [comment]
I'm getting the same issue as well.
@fcturner might we convince you to switch to nextcloud maybe :p looks like it's the future.

**Comment by fcturner** [comment]
I'm not so sure about that :wink:
I do like their dedication, but I don't like their style and policy. We'll see.

**Comment by moretocome** [comment]
Same issue on NC11.0 stable. All passwords are in the recycle bin. Password sync with Firefox extension still seems to work though.

I second the emotion by @tapir on NC vs OC. NC seems superior. The _Passwords_ app is by far my favorite app in the NC/OC app store though, and I am grateful for you work! Let me know if you need logs or similar.

Merry christmas @fcturner and everyone else!

**Comment by fcturner** [comment]
I am now running ownCloud 9 alongside NextCloud 11. #bugsqueezingstarted

I will do my best to keep supporting both platforms.

**Comment by seraphyn** [comment]
@fcturner Thank you and merry christmas

**Comment by tapir** [comment]
You're the man @fcturner 💯

**Comment by Luticus** [comment]
@fcturner That's awesome that you're trying to support both, just thought I'd give you a heads up that the latest beta breaks things more.  All passwords end up in the trash.  rolling back to v19 fixes that.  Also there seems to be a problem with the settings in the personal page.  The settings are blanked out for me.  In the debug console I see this error:

Uncaught TypeError: Cannot read property 'toLowerCase' of undefined
    at HTMLDocument.<anonymous> (settings.js?v=3127603…:114)
    at j (jquery.min.js?v=3127603…:2)
    at Object.fireWith [as resolveWith] (jquery.min.js?v=3127603…:2)
    at Function.ready (jquery.min.js?v=3127603…:2)
    at HTMLDocument.I (jquery.min.js?v=3127603…:2)

in version 19.

**Comment by fcturner** [comment]
Guys, this is a really, really nasty bug. It's a cryptographical error. Most depressing part: at the moment I don't know how to solve this. 

I'll try to explain this.
For verification of the encrypted passwords in your databases, I use [_keyed-hash message authentication code_ (HMAC)](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) which is a specific type of [_message authentication code_ (MAC)](https://en.wikipedia.org/wiki/Message_authentication_code). It's one my my most important functions, because it's a perfect test to validate a decryption attempt.

Problem is, the verification of this code fails on NextCloud 11 (**not** on earlier versions and **not** on ownCloud). 

This part of my code is PHP only, so I thought only the `mcrypt` library was needed to implement this. That's why `mcrypt` is a dependency of my app.

> @moretocome Same issue on NC11.0 stable. All passwords are in the recycle bin. Password sync with Firefox extension still seems to work though.

Even more difficult to relate.

**One question though:** I'm running v20b6 on NC11. My passwords aren't even avalible in the trash, nothing decrypts. How is that for you? Are the passwords in the trash being decrypted? Can you actually view those passwords?

I'm working my butt off to get this fixed before Christmas 😆

**Comment by tapir** [comment]
I can most definitely decrypt the passwords in trash. Even If I try to restore them they end up in trash next time I open passwords app.

So I'm really suspicious about this being about encryption/decryption.

Hey really thanks for doing this. It's much appreciated.

**Comment by pafcioooo** [comment]
I can see all passwords in recycle bin (trash).

**Comment by jiml8** [comment]
> Problem is, the verification of this code fails on NextCloud 11 (not on earlier versions and not on ownCloud).

This part of my code is PHP only, so I thought only the mcrypt library was needed to implement this. That's why mcrypt is a dependency of my app.

I regret moving to NC11.  NC10 was stable and worked well, but NC11 is inadequately tested.  I have found three obvious bugs in the core of NC11 that would not be there if they had tested adequately before releasing, and it would appear this is a fourth bug.

For reference, I am running NC11 on Nas4Free which is based on FreeBSD 11 (PHP 7), and my passwords are not in the trash.  They are where they belong, and they do decrypt.  I also have successfully added new accounts/passwords since the move to NC11.   

As you may recall, I am getting the json error in requesters, and apparently it is connected with password sharing in my instance. 

Thus, we seem to have a platform-based difference in the symptoms; I infer that the other people who have posted on this thread are running Linux.  Also, I first deployed passwords on Owncloud 8 with Nas4free 9 (FreeBSD 9) and have upgraded that server through FreeBSD 10 and to the current FreeBSD 11.  I don't know if the relatively ancient age of my passwords database would be relevant in any fashion.

> Problem is, the verification of this code fails on NextCloud 11 (not on earlier versions and not on ownCloud).

Please explain what you mean by "the verification of this code fails".  There is some parser complaint?  Or are you encountering some difficulty decrypting the HMAC codes?

**Comment by moretocome** [comment]
> @fcturner One question though: I'm running v20b6 on NC11. My passwords aren't even available in the trash, nothing decrypts. How is that for you? Are the passwords in the trash being decrypted? Can you actually view those passwords?

Yes with v20b6 all passwords ended up in the bin, but they did decrypt, and I was also able to make a plain text backup.

As suggested by @Luticus downgrading to v19 has fixed that. Now NC only throws the to errors referenced on @pafcioooo's screen shots, but all passwords are in their right place, and I am able to view, edit and create new.

Your efforts are appreciated @fcturner! Save some time for the loved ones as well :)

**Comment by jiml8** [comment]
I should emphasize that I have not deployed v20 of passwords.  Mine is v19.

**Comment by fcturner** [comment]
> @jiml8 Please explain what you mean by "the verification of this code fails". There is some parser complaint? Or are you encountering some difficulty decrypting the HMAC codes?

Alright, this will be a bit technical. Being sort of a cryptologist I find it hard to explain this easily. But I'll try my best.

_You can view the decryption function_ [_here in `passwordservice.php`_](https://github.com/fcturner/passwords/blob/master/service/passwordservice.php#L700-L732).

In fact, the input of the decryption function is `$data_hex`. That's the value you can see in your database fields, really long hexadecimal values. This value is converted to a binary value by using the PHP function [`hex2bin`](http://php.net/manual/en/function.hex2bin.php), which is supplied within the function if the function doesn't exists on your server because it is not available on PHP versions lower than 5.4.0. Then, the password salt (`$salt`), the actual encrypted part (`$enc`) and the MAC (`$mac`) are extracted from that. The HMAC hashes of those variables are compared and matched using the PHP function [`hash_hmac`](http://php.net/manual/en/function.hash-hmac.php) and my own function `hash_equals`. The latter you can review here:
```php
function hash_equals($a, $b) {
	$key = mcrypt_create_iv(128, MCRYPT_DEV_URANDOM);
	return hash_hmac('sha512', $a, $key) === hash_hmac('sha512', $b, $key);
}
```

These values should be the same. Decryption fails when they aren't.

**Comment by fcturner** [comment]
Adding new passwords works on my NC11 instance by the way, albeit being them thrown in the trash immediately. But all other passwords aren't decrypted.

**Comment by fcturner** [comment]
I created a new branch in #301, would you like to test it? It doesn't solve anything, but writes errors to the log.

**Comment by fcturner** [comment]
THINKING ABOUT THE SALT VARIABLE:

Are you all aware that the encryption key is instance specific? That's part of the security I built in. So the 'passwordsalt' in config.php needs to be the same as the one that was set in instance where the password was created in.

**Comment by tapir** [comment]
I copied over my config.php so it should be fine right?

**Comment by jiml8** [comment]
I asked you the question because I develop.  Presently I am one of the lead developers on a commercial security product that does use encryption, and I do program crypto (though usually in C) - and I also program PHP.

I note in the code you referenced to me that you are using the PHP list command.  Take a look at this:

"WARNING:  In PHP 5, list() assigns the values starting with the right-most parameter. In PHP 7, list() starts with the left-most parameter.

If you are using plain variables, you don't have to worry about this. But if you are using arrays with indices you usually expect the order of the indices in the array the same you wrote in the list() from left to right, which is not the case in PHP 5, as it's assigned in the reverse order.

Generally speaking, it is advisable to avoid relying on a specific order of operation, as this may change again in the future. "

http://php.net/manual/en/function.list.php

**Comment by jiml8** [comment]
I also notice this:

                //if (!hash_equals(hash_hmac('sha512', $enc, $macKey, true), $mac)) {
		if (!Encryption::hash_equals(hash_hmac('sha512', $enc, $macKey, true), $mac)) {
			 return false;

I am guessing the commented out line is from an earlier version of passwords.  I note that you are referring to hash_equals with the included class name, which should not be necessary since hash_equals is defined in the class, and the function decrypt is also in the class.  I am wondering why you did this?  If the program failed unless you did that, then perhaps there is a problem with your class, or a namespace collision, or some such.

**Comment by bschelst** [comment]
my entries are also going to trash bin, as mentioned by lucashalbert

**Comment by SaschaDe** [comment]
Hi,
I update the file passwordservice.php from fcturner.
All my Passwords going to trash bin. But I can use the passwords from the trash bin without problems.

Logfile:

`Decryption of id 9 of user 'xxx' failed: the HMAC hashes of the MAC and the hash of MAC does not match with hash of HMAC`

The message repeats for all existing passwords.
When I click the "refresh" button from a password then the entry moved from the trash bin to active passwords for this session.

**Comment by aktavor69** [comment]
Same as @SaschaDe

**Comment by Luticus** [comment]
I'm not sure if it's just me or only my account on my cloud or whatever, but I did NOT roll forward to passwords 20 beta 6 and instead stayed on passwords 19, then I replaced my passwordservice.php file with the one fcturner provided and like magic everything seems fine now.  I still need to skim my logs and have other users report in to let me know, but at least for me everything works with no errors and no items are going to the trash (that seems to be caused by something specific in passwords 20 beta 6. If you downgrade back to 19 (i had previously upgraded) it will automatically move all your passwords back from the trashcan, at least it did for me.  If anyone is going to try this please do so only after exporting the passwords of all users to a backup file!  I'll let you know if anything changes.

**Comment by Luticus** [comment]
Things are still broken, the personal passwords settings are broken and so are the settings in the passwords admin panel.  Changing the file just masks the errors in the app.

**Comment by Luticus** [comment]
Also I've noticed that importing passwords from a csv is broken with the new passwordservice.php file.

**Comment by mlendvay** [comment]
Experiencing the same issue. The modified passwordservice.php stops the JSON errors, but when creating a password, a completely empty message popup is displayed, and the password is never created.

**Comment by alnixon** [comment]
I've added pull request #313 for a fix that worked for me.  Unlike others discussing this issue the only problem I was experiencing was the popups, all other functionality is working.  My passwords didn't move to the bin and I don't have a problem creating passwords.
I am running Nextcloud 11.0 from the stable releases.  I have release 19 of Passwords (from the tag, no additional commits).

**Comment by duh-nm** [comment]
Maybe its just me, and while this fixed the JSON errors, i still cant get any passwords back out of the trash. anyone else have any luck?

**Comment by Luticus** [comment]
Are you on 20 beta 6 or still on 19?  For me when I downgraded from 20 back to 19 my passwords were out of the trashcan and back in their normal place.  I'd recommend ya back everything up before you downgrade though.  I was able to with no visible problems but who knows with a downgrade.

**Comment by duh-nm** [comment]
i was up on beta7. i just dropped down to 19, and they are out of the trash, with the json errors back. you were running the updated php from the commit on 19?

**Comment by Luticus** [comment]
I was, I switched that back because that php file left me unable to save new passwords or import anything, although it did squelch the errors.  the 313 fix and passwords 19 is all i'm running now and while it doesn't seem to be perfect (there are still dialogue boxes that are off center on the screen), it does get rid of the errors as far as I can tell.

**Comment by marius-wieschollek** [comment]
I found out how to fix the issue with all passwords being in the trash:
in the passwords.js in line 1640 replace `if (row.deleted) {` with `if (row.deleted == "true") {`.
Works on my Nextcloud 11 with Firefox and Vivaldi. (Make sure to clear browser cache)

**Comment by pafcioooo** [comment]
@marius-wieschollek, that works for me too:-)

**Comment by jeweet** [comment]
Update to the latest git version fixed the json errors.
@marius-wieschollek: your fix indeed prevents the passwords being in the trash after login.

I'm on Nextcloud 11
Thanks for the fix!

**Comment by duh-nm** [comment]
Worked for me too. Rather weird.

**Comment by moretocome** [comment]
I can confirm beta7 fixes the json error, and the edit from @marius-wieschollek fixes the recycle bin problem.

In js/passwords.js i changed line 1640 from:
`	if (row.deleted || row.deleted == "true") {`
to:
`	if (row.deleted && row.deleted == "true") {`
though. All passwords are where they are supposed to be, and I can edit and create new.

I still had to remove the COLLATE argument to fix #308 

Unfortunately now my ff-extension is back to reporting "0 logins in database", but it seems the author of the extension has already committed a [fix](https://github.com/eglia/ff-oc-passwords/commit/59f71eb0f2d74fe12407148face66ad0f3ac9207).

**Comment by pafcioooo** [comment]
BTW, @marius-wieschollek change, which is part of a full solution to this issue is not in a source code. Though I found a separate issue #314 concerning this.

**Comment by hobas** [comment]
@moretocome , this did the trick on beta-7 and Nextcloud-11.0.2. All passwords are back again. Well done!
